[web-animations-2] Spec for event-based animation triggers

szager-chromium opened the pull request:
Specification for [event-based animation triggers](https://github.com/szager-chromium/declarative-interactions)

tidoust committed:
[css-color-hdr-1] Use definition of `color()` from css-color-5

Issue #11954 noted that the spec does not really extend the definition of the
`color()` function but rather extends the definition of one of the types that
gets used in the definition of `color()`. A previous update dropped duplicated
constructs accordingly, but the duplication of the definition itself remained.

Duplication is confusing for tools and readers because there's no longer any
direct linnk between the definition of the `color()` function and its syntax.

This replaces the definition with a reference to the actual definition in
`css-color-5`, allowing readers to access the actual function syntax. Note
the "link default" should only be needed on a temporary basis (and it is needed
precisely because the cross-references database currently contains more than
one definition of the `color()` function).

szager-chromium committed:
[web-animations-2] Spec for event-based animation triggers

szager-chromium committed:
bikeshed previews

frivoal committed:
[css-ui-4] Fix typo

see https://github.com/w3c/csswg-drafts/issues/12194

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2366,15 +2367,29 @@ Animation Triggers {#triggers}
+and an <dfn export >animation trigger name</dfn>.
+Changes to animation playback occur when an {{Event!!interface}} is dispatched
+with {{Event/type}} matching the trigger's associated [=animation trigger/event type=]
+and {{Event/target}} which is an {{Element}} whose 'animation-trigger-name'
```
So to be clear only events directly targeting the element, not those targeting descendants which would bubble to the element trigger the animation?

szager-chromium replied:
Yes precisely; no bubbling.

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2460,13 +2480,42 @@ given values of <var ignore=''>behavior</var>:
+
+<dl>
+  <dt>''animation-trigger-type/once''</dt>
+  <dd>If [=play state=] is [=play state/finished=], do nothing.
```
Do event-based animation triggers have the same behavior as timeline-based triggers before the trigger occurs? If so, and per the direction we're leaning in #12119 we should only play the animation if it's in the paused waiting for trigger state. @DavMila

szager-chromium replied:
tl;dr on that issue (I read about half of it). My understanding of animation state is still based on [play states](https://drafts.csswg.org/web-animations-1/#play-states) and I'm not sure how triggers change that model, but my assumptions are:

- if the animation is `idle`, `play()` will start the animation
- if the animation is `running`, `play()` will be a no-op
- if the animation is `paused`, `play()` will transition it to `running`
- if the animation is `finished`, we shouldn't do anything

That's how I came up with this spec text. If I'm missing a concept here, please let me know.

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2460,13 +2480,42 @@ given values of <var ignore=''>behavior</var>:
+
+  <dt>''animation-trigger-type/repeat''</dt>
+  <dd>First, if [=play state=] is [=play state/running=] or [=play state/paused=]
+      perform the actions specified for the {{Animation/cancel}} function.
```
You want to restart the animation if it was previously in progress? This is very different from the behavior we have for scroll triggered animations where if the animation was already playing it just continues playing. Is it intentional that this is different? Does it make sense for events or would there be some use cases where we want the playing animation from the earlier event to finish?

szager-chromium replied:
Unlike event-based triggers, timeline-based triggers are essentially a toggle. My mental model for this is that the behavior for an event-based trigger is similar to a matched pair of `primary->inverse` and `inverse->primary` trigger events for a timeline-based trigger. I may be misreading the spec language for timeline-based trigger behavior here, because it's not expressed in terms of WAAPI functions:

```
If state is primary, the effect is to trigger the associated animation.
If state is inverse, the effect is to reset the associated animation effect back to its before phase
and its associated animation’s start time to zero.
```

Is the `inverse` behavior not roughly equivalent to `cancel()`?

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2460,13 +2480,42 @@ given values of <var ignore=''>behavior</var>:
+  </dd>
+
+  <dt>''animation-trigger-type/alternate''</dt>
+  <dd>Perform the actions specified for the {{Animation/reverse}} function.</dd>
```
If the play state is paused though it should start playing, right?

szager-chromium replied:
The [spec](https://drafts.csswg.org/web-animations-1/#dom-animation-reverse) for `reverse()` says that it will run the animation.

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2460,13 +2480,42 @@ given values of <var ignore=''>behavior</var>:
 Issue: Do we need a proper definition for "behavior types primary/inverse"?
 
-### Animation Trigger Active Interval ### {#trigger-active-interval}
+#### Event-Based Animation Trigger Behavior #### {#event-trigger-behaviors}
```
Rather than mix the timeline and event based trigger sections it seems like we should have a different top-level section for timeline vs event-based triggers, e.g.

* Animation Triggers
  * Overview
  * Event-based
    * Behavior
  * Timeline-based
    * Behavior
    * Range
    * State
    * etc
  * Updating

szager-chromium replied:
Done.

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2460,13 +2480,42 @@ given values of <var ignore=''>behavior</var>:
+
+For [=event-based animation triggers=] changes to animation playback happen when a
+qualifying event occurs, according to the trigger's [=animation trigger behavior=]
+and the current [=play state=] of the associated [=animation=]:
```
I suspect that the result of #12119 may be that we can share more of this functionality, though I suppose it would be fine to merge the two behavior sections afterwards. Let's try to align though on similar behaviors as much as it makes sense. @DavMila

szager-chromium replied:
Agreed!

DavMila replied:
I think this partly relates to https://github.com/w3c/csswg-drafts/issues/11914.

If we specify that playback methods disable triggers, then the trigger doesn't need to take Animation.playState into account and we can write our expectations for a `once` trigger, for example, (event- or timeline- based) just in terms of the trigger's state, i.e. a `once` trigger plays its animation when it transitions from `idle` to `primary` (which can only happen once). 

If they don't disable  triggers, then we might want triggers to be a little smarter and perhaps it makes sense to factor in the animation's playState. I think the logic would still be the same for event-based and timeline-based.

szager-chromium replied:
My intention here is that event-based triggers do not have a "trigger state". I don't even think it makes sense to have an "idle" pre-period for event-based triggers, it's just a bad conceptual fit.

I'm also not sure it makes sense to disable an event-based trigger whenever a WAAPI method is called on an animation. I need to mull that over a bit, but my first instinct is that it would be confusing/surprising.

flackr commented on web-animations-2/Overview.bs:
```diff
@@ -2366,15 +2367,29 @@ Animation Triggers {#triggers}
+Changes to animation playback occur when an {{Event!!interface}} is dispatched
+with {{Event/type}} matching the trigger's associated [=animation trigger/event type=]
+and {{Event/target}} which is an {{Element}} whose 'animation-trigger-name'
+property matches the trigger's associated [=animation trigger name=].
```
In the web-animations spec, triggers should not be associated by name. If you're programmatically constructing a trigger you can construct a trigger for a particular element, and add the animation to that trigger, all without using names which could accidentally result in triggering unrelated animations.

flackr commented on css-animations-2/Overview.bs:
```diff
@@ -962,6 +962,52 @@ The values of 'animation-trigger-exit-range-start' have the following meaning:
                         it defaults to 100%.
 </dl>
 
+## The 'animation-trigger-name' property ## {#animation-trigger-name}
```
I think this needs some thought. Do you have some examples of how this would be used. We should model this after named timelines, see in particular:
* [Named scroll progress timelines](https://drafts.csswg.org/scroll-animations-1/#scroll-timelines-named)
* [Named view progress timelines](https://drafts.csswg.org/scroll-animations-1/#view-timelines-named)
* [Named timeline scoping and lookup](https://drafts.csswg.org/scroll-animations-1/#timeline-scoping)

The name lookup you've defined here sounds global - it should be scoped similar to timelines. I think what you want is for a named trigger to be created by these properties and then possible to lookup by the animation-trigger property.

Further, to avoid confusion, we should keep the `animation-*` properties only for properties which are directly associated with their corresponding `animation-name` in the animation list and use a different property prefix to establish event triggers, e.g. `event-trigger-*`

@ydaniv @DavMila @birtles perhaps timeline triggers should also be set up using a separate namespace of properties and then associated with animations by scoped name?

noamr committed:
[css-values-4] [css-cascade-4] [css-color-5] [css-fonts-4] [css-images-4] [css-shapes-2] Clean up fetching (#12261)

* [css-vlaues-4] [css-cascade-4] [css-color-5] [css-fonts-4] [css-images-4] [css-shapes-2] Clean up fetching

This fixes a few cases around fetching external URLs for style resources:
- The base URL is the sheet's base URL if exists, otherwise the sheet's location, otherwise the document
  base URL.
- The algorithm can take either a rule or declaration block, and derive the correct stylesheet and base
  URL from that.
  Updated the calling sites to take that onto account.

- Removed redundant URL parsing in "@import".

Closes #12065
Closes #12068
Closes #12086
Closes #12147

* Extract the computed absolute URL to a separate algo

* A few fixes

* nit

cdoublev committed:
[css-values-5] Make color-mix() accept 1+ arguments #11530 (#12265)

Oversight in 85cd6d2.

noamr committed:
[css-borders-4] Define border/shadow rendering for `corner-shape` (#12175)

* [css-borders-4] Define  border/shadow rendering

* Add note

* Update chart

* Overhaul the algo

* nit

andreubotella committed:
[css-overflow-4] Draft spec for `continue: collapse` (#7708) (#10816)

* [css-overflow-4] Draft spec for `continue: collapse` (#7708)

* Clarification about float clearance in the automatic block size

* Update to only allow clamping after a line

* Apply suggestions from code review

Co-authored-by: Florian Rivoal <git@florian.rivoal.net>

* Update css-overflow-4/Overview.bs

Co-authored-by: Florian Rivoal <git@florian.rivoal.net>

* Update css-overflow-4/Overview.bs

* Update css-overflow-4/Overview.bs

* Update css-overflow-4/Overview.bs

* Update css-overflow-4/Overview.bs

* Update css-overflow-4/Overview.bs

* Update for #10439

* Update for #11962

* Update for #10868

---------

Co-authored-by: Florian Rivoal <git@florian.rivoal.net>
Co-authored-by: Tab Atkins Jr. <jackalmage@gmail.com>

andruud committed:
[css-mixins-1] Consistently return a syntax string (#12245)

Perhaps these functions should indeed return a serialized <css-type>
(#11908), but until then they should always return a valid syntax string.

Co-authored-by: Anders Hartvoll Ruud <andruud@google.com>

noamr committed:
[css-borders-4] Constrain radii for concave opposite corners (#12272)

* [css-borders-4] Constrain radii for concave opposite corners

If the hulls of opposite corners overlap, find a scale factor that would prevent the overlap, and apply to all radii.

Resolution: https://github.com/w3c/csswg-drafts/issues/12098#issuecomment-2898621830
Closes #12098

* Fix hull computation for convex/concave

andruud committed:
[css-mixins-1] Don't erase the result type (#12259)

The "result" registration got caught in the blast radius when resetting
the types for the local variables.

Co-authored-by: Anders Hartvoll Ruud <andruud@google.com>

svgeesus committed:
[css-shapes-1][editorial] Add WPT annotations

szager-chromium committed:
in progress

svgeesus committed:
[css-shapes-1][editorial] Updated changes

svgeesus committed:
[css-color-4] Interpolation between legacy sRGB colors is (once again) in sRGB space, for compatibility. #7949

noamr committed:
Move shape() to css-shapes-1 (#12279)

noamr committed:
[css-shapes-1] Minor fixes for WD (#12284)

noamr committed:
[css-values-4] Fetching: use declaration block instead of declaration (#12286)

* [css-values-4] Fetching: use declaration block instead of declaration

* nit

svgeesus committed:
[css-color-4][editorial] linkable note

tabatkins committed:
[css-borders-4] Fix image

tabatkins committed:
[css-borders-4] Rearrange and editorial rewrite the non-algo parts of the corners section.

tabatkins committed:
[css-borders-4] fix caption

noamr committed:
[css-borders-4] Add a general note about outline bordering and corner-shape (#12278)

Resolution:
https://github.com/w3c/csswg-drafts/issues/12194#issuecomment-2917026075

Closes #12194

szager-chromium committed:
Merge branch 'css-animations-2-bikeshed' of https://github.com/szager-chromium/csswg-drafts into animation-trigger-event

szager-chromium committed:
[css-animations-2] Fix bikeshed errors (#12253)

Editorial linking error fixes.

tabatkins committed:
[css-borders-4][editorial] Simplify the SVG markup

tabatkins committed:
[css-transforms-1][editorial] Export 'identity transform' for convenience.

fantasai committed:
[css-values-5] Replace mix() functions with mix() + interpolate() #6245

See https://github.com/w3c/csswg-drafts/issues/6245#issuecomment-2661545284
and https://lists.w3.org/Archives/Public/www-style/2025Feb/0020.html

frivoal committed:
[css-overflow-4] Adjust example 6 to also work with continue:collapse

What it illustrating is equally applicable to continue:collapse and
continue:discard, but it was only written in terms of the latter.
Tweak the wording so that it applies to both.

frivoal committed:
[css-overflow-4] Tweak note on line-clamp and webkit-box

Since -webkit-legacy is now a possible value of line-clamp,
the text in the note needed a bit of refinement to clarify that it is
talking about the case where that value is not set.

frivoal committed:
[css-overflow-4] phrasing tweak

Adjust the order of the text to match the order of the grammar, in order
to avoid suggesting a different ordering.

See #8742

cdoublev committed:
[css-borders-4] Typo (#12297)

tidoust committed:
[css-shapes-2] Fix typo in link to css-shapes-1 (#12295)

andruud committed:
[css-mixins-1] Check default values against the parameter type (#12243)

Co-authored-by: Anders Hartvoll Ruud <andruud@google.com>

dontcallmedom committed:
Fix link to MathML section (#12242)

tabatkins committed:
[css-values-5] Fix normalization steps. #6245

frivoal committed:
[css-overflow-4][editorial] Minor phrasing tweak (#12293)

*Something* happens to the excess content (it overflows), so I find it
more natural to say that the behavior is the same as the initial value,
rather than to say there is no effect.

frivoal committed:
[css-overflow-4] phrasing tweak in setting clamp points (#12294)

Adjust the phrasing of the sentence describing the case when you cannot
set a clamp point. In the original phrasing:
* it wasn't clear what "it" referred to
* the reference to other clamp points without a qualifier like
  "possible" or "potential" made the logic seem circular.

chrishtr committed:
Update title

cdoublev committed:
[css-values-5] Typo (#12305)

* [css-values-5] Typo

* [css-values-5] Prevent re-ordering *-interpolate() first argument

* [css-syntax-5] Typo

cdoublev committed:
[css-values-5] Make the progress() result consistent with its arguments (#12309)

This was accidentally removed in 640e1c0f4819f3eb2bf076c37ed233059d3327e9.

szager-chromium committed:
in progress

szager-chromium committed:
Merge branch 'main' of https://github.com/w3c/csswg-drafts into animation-trigger-event

szager-chromium committed:
nominally complete

szager-chromium committed:
Delete html output

szager-chromium committed:
fix whitespace
